packages/vite-plugin-cloudflare/src/cloudflare-environment.ts

```diff
@@ -182,13 +193,43 @@ export function createCloudflareEnvironmentOptions(
 					".cts",
 					".ctx",
 				],
+				define,
 			},
 		},
-		// if nodeCompat is enabled then let's keep the real process.env so that workerd can manipulate it
-		keepProcessEnv: isNodeCompat(workerConfig),
+		// We manually set `process.env` replacements using `define`
+		keepProcessEnv: true,
 	};
 }
 
+/**
+ * Gets `process.env` replacement values.
+ * `process.env.NODE_ENV` is always replaced.
+ * `process.env` is replaced with an empty object if `nodejs_compat` is not enabled
+ * @param workerConfig - the Worker config (used to determine if `nodejs_compat` is enabled)
+ * @param mode - the Vite mode
+ * @returns replacement values
+ */
+function getProcessEnvReplacements(
+	workerConfig: WorkerConfig,
+	mode: string
+): Record<string, string> {
+	const nodeEnv = process.env.NODE_ENV || mode;
+	const nodeEnvReplacements = {
+		"process.env.NODE_ENV": JSON.stringify(nodeEnv),
+		"global.process.env.NODE_ENV": JSON.stringify(nodeEnv),
+		"globalThis.process.env.NODE_ENV": JSON.stringify(nodeEnv),
+	};
+
+	return isNodeCompat(workerConfig)
+		? nodeEnvReplacements
+		: {
+				...nodeEnvReplacements,
+				"process.env": "{}",
+				"global.process.env": "{}",
+				"globalThis.process.env": "{}",
```

Comment:
could you explain the logic here? 🙂 

Why setting the `process.env.NODE_ENV` but then setting all the rest to `{}`, is it really necessary for force `process.env` to be `{}`? (for example I wonder, if some user code / library were to polyfill `process.env` could this break them? 🤔)

Reply:
This copies the logic from [here](https://github.com/vitejs/vite/blob/d52cf887fddcfe9e19b02d20de302fae4bbffd1f/packages/vite/src/node/plugins/define.ts#L24-L26). I think the rationale is that it's safer for `process.env` to be an empty object than left as is in a Worker without `nodejs_compat` enabled. This is because if code tries to access a property on it then it will get `undefined` rather than throwing `Cannot read properties of undefined`.